tests/MatchBot/Application/Matching/OptimisticRedisAdapterTest.php

```diff
@@ -0,0 +1,171 @@
+        });
+    }
+
+    public function testItSubtractsAmountForFunding(): void
+    {
+        $this->sut->runTransactionally(function () {
+            $funding = new CampaignFunding();
+            $funding->setAmountAvailable('50');
+            $amountToSubtract = "10.10";
+
+            $this->entityManagerProphecy->persist($funding)->shouldBeCalled();
+            $fundBalanceReturned = $this->sut->subtractAmount($funding, $amountToSubtract);
+
+            \assert(50 - 10.10 === 39.9);
+            $this->assertSame('39.90', $this->sut->getAmountAvailable($funding));
+            $this->assertSame('39.90', $fundBalanceReturned);
+        });
+    }
+
+    public function testItReleasesFundsInCaseOfRaceCondition(): void
+    {
+        $this->sut->runTransactionally(function () {
+                $funding = new CampaignFunding();
+                $funding->setAmountAvailable('50');
+                $amountToSubtract = "30";
+
+            $this->entityManagerProphecy->persist($funding)->shouldBeCalled();
+            $this->sut->subtractAmount($funding, $amountToSubtract);
+                try {
+                    $this->sut->subtractAmount($funding, $amountToSubtract); // this second subtraction will take the fund negative in redis temporarily, but our Adapter will add back the 30 just subtracted.
+                    $this->fail("should have thrown exception on attempt to allocate more than available");
+                } catch (LessThanRequestedAllocatedException $exception){
+                    $this->assertStringContainsString("Less than requested was allocated", $exception->getMessage());
+                }
+
+                $this->assertSame('0.00', $funding->getAmountAvailable());
+
+                // this seems like it could be a bug as the amount in storage is negative after the exception was caught?
+                $this->assertSame('0.00', $this->sut->getAmountAvailable($funding));
+        });
+    }
+
+
+    public function testItBailsOutAndReleasesFundsIfRetryingDoesntWorkDueToConcurrentRequests(): void
+    {
+        // let's assume another thread is causing the funds to reduce by 30 just
+        // after each time we increase it by 30.
+        $this->storage->setPreIncrCallBack(function (string $key) {
+            return $this->storage->decrBy($key, 30);
+        });
+
+        $this->sut->runTransactionally(function () {
+            $funding = new CampaignFunding();
+            $funding->setAmountAvailable('50');
+            $amountToSubtract = "30";
+
+            $this->sut->subtractAmount($funding, $amountToSubtract);
+
+            $this->expectException(TerminalLockException::class);
+            // todo - work out where the 1880 figure here comes from. Message below is just pasted in from
+            // result of running the test.
+            $this->expectExceptionMessage("Fund  balance sub-zero after 6 attempts. Releasing final 1880 'cents'");
```

Comment:
Is it worth fund having a (fixed?) fund ID to make it clearer how this message is built in the test + that it's not a typo?

Reply:
Done in f139315